[abseil] Configure abseil to use std:: types when feature cxx17 is enabled

cneumann opened the pull request:
Patches `absl/base/options.h` type selection macros to use either value
0 (when feature `cxx17` is not enabled) to ensure that always absl:: types are used. This was already
done by the pre-existing `fix-lnk2019-error.patch` in the port.
When feature `cxx17` is enabled instead value 1 is used to force abseil to use
the std:: types.

Abseil can use `std::any`, `std::optional`, `std::string_view`, and `std::variant`
or its own internal replacement types. Which ones are used
is configured by setting macros in `absl/base/options.h` to values 0/1/2
where 0 means always use `absl::` types, 1 is always use `std::` types (and
require a C++17 library), 2 detect if the standard library supports the
types and use them if available (see also comments in `absl/base/options.h`).

cneumann committed:
[abseil] Use std:: types when feature cxx17 is enabled

Adds a patch changing macros in absl/base/options.h to
always use the std:: namespace types instead of the
absl:: namespace replacements (for any, optional,
string_view, variant).
The upstream version of options.h uses compiler/library
feature detection to decide this, but patch
fix-lnk2019-error.patch hard codes use of absl:: types,
thus rendering setting CMAKE_CXX_STANDARD to 17 in the
port file ineffective. Since auto detection is problematic
from an ABI point of view (see comments in
absl/base/options.h for details), this applies an
alternate patch for fix-lnk2019-error.patch when feature
cxx17 is enabled.

LilyWangL commented:
Thanks for your PR! Can you please correct the Version field in the CONTROL file? You can get more information from [maintainer-guide.md](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md).

cneumann commented:
Ah, sorry for missing that. I've updated the version and also updated the commit message to include the port tag and hopefully be a bit clearer about what is being fixed and how.

LilyWangL commented:
/azp run

cneumann commented:
Hmm, I'm having trouble understanding the cause for the failures reported above. For example looking at the details of `vcpkg-osx-PR-test` it looks like the port that fails is `mlpack`, but as far as I can tell `abseil` is not a direct nor indirect dependency of `mlpack` - how did my change manage to break it? Any pointers are appreciated :)

LilyWangL commented:
> Hmm, I'm having trouble understanding the cause for the failures reported above. For example looking at the details of `vcpkg-osx-PR-test` it looks like the port that fails is `mlpack`, but as far as I can tell `abseil` is not a direct nor indirect dependency of `mlpack` - how did my change manage to break it? Any pointers are appreciated :)

It's not related with this PR, this error will be fixed in PR #11057.

LilyWangL commented:
/azp run

strega-nil commented:
This is great, thanks @cneumann :)
